crml/nft/src/lib.rs

```diff
@@ -0,0 +1,651 @@
+		NotForAuction,
+		/// Cannot operate on a listed NFT
+		TokenListingProtection,
+		/// Internal error during payment
+		InternalPayment,
+		/// Total royalties would exceed 100% of sale
+		RoyaltiesOvercommitment,
+		/// Auction bid was lower than reserve or current highest bid
+		BidTooLow
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Nft {
+		/// Map from collection to owner address
+		pub CollectionOwner get(fn collection_owner): map hasher(blake2_128_concat) CollectionId => Option<T::AccountId>;
+		/// Map from collection to its schema definition
+		pub CollectionSchema get(fn collection_schema): map hasher(blake2_128_concat) CollectionId => Option<NFTSchema>;
+		/// Map from collection to it's defacto royalty scheme
+		pub CollectionRoyalties get(fn collection_royalties): map hasher(blake2_128_concat) CollectionId => Option<RoyaltiesSchedule<T::AccountId>>;
+		/// Map from a token to it's royalty scheme
+		pub TokenRoyalties get(fn token_royalties): double_map hasher(blake2_128_concat) CollectionId, hasher(twox_64_concat) T::TokenId => Option<RoyaltiesSchedule<T::AccountId>>;
+		/// Map from (collection, token) to it's attributes (as defined by schema)
+		pub Tokens get(fn tokens): double_map hasher(blake2_128_concat) CollectionId, hasher(twox_64_concat) T::TokenId => Vec<NFTAttributeValue>;
+		/// The next available token Id for an NFT collection
+		pub NextTokenId get(fn next_token_id): map hasher(twox_64_concat) CollectionId => T::TokenId;
+		/// The total amount of an NFT collection in existence
+		/// Map from (collection, token) to it's owner
+		pub TokenOwner get(fn token_owner): double_map hasher(blake2_128_concat) CollectionId, hasher(blake2_128_concat) T::TokenId => T::AccountId;
+		/// Map from (collection, account) to the account owned tokens of that collection
+		pub CollectedTokens get(fn collected_tokens): double_map hasher(blake2_128_concat) CollectionId, hasher(blake2_128_concat) T::AccountId => Vec<T::TokenId>;
+		/// The total amount of an NFT collection in existence
+		pub TokenIssuance get(fn token_issuance): map hasher(blake2_128_concat) CollectionId => T::TokenId;
+		/// The total amount of an NFT collection burned
+		pub TokensBurnt get(fn tokens_burnt): map hasher(blake2_128_concat) CollectionId => T::TokenId;
+		/// NFT sale/auction listings. keyed by collection id and token id
+		pub Listings get(fn listings): double_map hasher(blake2_128_concat) CollectionId, hasher(twox_64_concat) T::TokenId => Option<Listing<T>>;
+		/// Winning bids on open listings. keyed by collection id and token id
+		pub ListingWinningBid get(fn listing_winning_bid): double_map hasher(blake2_128_concat) CollectionId, hasher(twox_64_concat) T::TokenId => Option<(T::AccountId, Balance)>;
+		/// Map from block numbers to listings scheduled to close
+		pub ListingEndSchedule get(fn listing_end_schedule): map hasher(twox_64_concat) T::BlockNumber => Vec<(CollectionId, T::TokenId)>;
```

Comment:
The fact that we cannot have the same pair of (collection id - token id) repeated for a specific block number suggests we better have BTreeSet instead of Vec here.

Reply:
I ended up changing this to a double map